### PR TITLE
feat(recovery): surface the attempt-preserve ref in defer notification and status

### LINKED ISSUE #333: Name the attempt-preserve ref in the defer notification, status, and --json
When a story is deferred after review-budget exhaustion, the engine parks the attempt's work on a git ref and then throws the ref away. Nothing the operator can see — the notification, `bmad-loop status`, and `status --json` — ever names it. The reporter of the umbrella field report recovered their work by running `git log --all` and guessing.

## Evidence (validated at `3c68153`)

**The ref is created, named well, and journaled:**

- Committed attempt work → branch `attempt-preserve/<run-id-slug>-<head8>` — `recovery_flow.py:317-324`, via `verify.preserve_commits` (`verify.py:266`), which *returns* the ref.
- Dirty worktree → `refs/attempt-preserve-dirty/<slug>-<baseline8>-<attempt>` — `recovery_flow.py:360`.
- Retention pruned to `scm.preserve_keep` (default 20).
- Recorded only as journal events `attempt-commits-preserved` / `attempt-worktree-preserved` — `recovery_flow.py:335-337`, `:374`.

**Then it is dropped:**

- `preserve_attempt_commits` returns `None` (`recovery_flow.py:292`) — the ref never reaches task or run state.
- No `src/` consumer of those journal events. Not in `diagnostics.py`, not in the TUI.
- Status text prints only `defer_reason` / `commit_sha` — `cli.py:1581-1585`.
- `--json` projection has no field — `documents.py:227-239`. `cmd_status` never opens `journal.jsonl`.

**The defer notification is also thinner than escalation's.** `_defer` *does* call `gates.notify` on both arms (`engine.py:3002` worktree, `:3022` in-place) — the message body is just the bare reason, with no action tail. Contrast escalation, which composes `f"{reason} — resolve, then \`bmad-loop resume {run_id}\`"` (`engine.py:3079-3084`). Prior art for a good recovery message already exists: `recovery_flow.py:409-459` manual-recovery instructions do "your work is at X, run `git -C <root> …`".

Zero test coverage pins that `_defer` / `_escalate` notify at all (`grep ATTENTION tests/` → gates unit tests + adapter budget-trip only), so this is regression-invisible.

## Fix outline

- New `StoryTask.preserve_ref: str | None` (`model.py` after `:176`; additive `to_dict`/`from_dict`, drop-in like `restore_patch`).
- Set it **inside `RecoveryFlow`** (its methods already receive the task): clear to `None` at the top of the auto-recover arm (`recovery_flow.py:165`), assign in `preserve_attempt_commits` (`:335`) and `preserve_attempt_worktree` (`:373`). Dirty (last writer) wins — the dirty snapshot is parented at attempt HEAD (`verify.py:446-449`), so it subsumes the commits branch and one uniform `git merge --ff-only <ref>` recovers either case.
- `_defer` in-place arm: notify message = reason + `` — attempt work parked at `<ref>`; recover with `git -C "<root>" merge --ff-only <ref>` `` (`engine.py:3022-3027`), plus a `preserve_ref=` field on the `story-deferred` journal entry (`:3021`).
- Isolated (worktree) arm: when `scm.keep_failed and task.branch`, append `` — failed work kept on branch `<branch>` ``. Do **not** set `preserve_ref` to the unit branch.
- Projections: `documents.py:227-239` adds `preserve_ref`; `cli.py:1581` text; `README.md:553`/`:465`; `docs/FEATURES.md:62`.
- Escalation path out of scope (no rollback there). No policy / `core.toml` changes. Not cleared on DONE — it stays as a mid-retry rollback breadcrumb.

## Tests

Extend `test_recovery_flow.py:269`/`:297`/`:336` plus a new both-families test (assert the dirty ref and `git merge-base --is-ancestor` subsumption); `test_engine.py:3204` defer test asserting ATTENTION content; `test_engine_worktree.py:297`/`:320` branch naming; `test_model.py` round-trip; `test_cli.py:865`/`:1076` json + text.

Two ablations required before trusting the suite (repo rule): (1) delete the `None`-clear → the stale-ref test must fail; (2) make the message tail unconditional → the bare-reason-when-no-ref test must fail.

Edge cases already analysed: replay re-defer re-parks under the same names (`branch -f`, attempt-keyed dirty ref); the pruned-ref caveat is documented, never validated at status time (status must not run git); message truncation rides the message, not the title.

## Acceptance

- The defer notification names the ref and gives a runnable `git merge --ff-only` command.
- `bmad-loop status` and `status --json` expose `preserve_ref` **additively** — `--json` schema stays v1.

Part of the 0.9.0 field-report umbrella (linked parent).

### PR DESCRIPTION
**Pull request by pbean**
Closes #333

A deferred story's rollback parks the attempt on a recovery ref, but the ref only ever reached `journal.jsonl` — `preserve_attempt_commits` dropped its return value, no `src/` consumer read the events, and the notification carried a bare reason. The reporter found their work with `git log --all`.

## What changed

- **`StoryTask.preserve_ref`** (`model.py`) — additive, round-trips, defaults `None` for pre-upgrade `state.json`.
- **`RecoveryFlow` sets it** — cleared at the top of every auto-recover arm, then assigned by `preserve_attempt_commits` (the `attempt-preserve/*` branch) and `preserve_attempt_worktree` (the `refs/attempt-preserve-dirty/*` snapshot). Dirty wins last: the snapshot is `commit-tree`'d parented at the attempt's HEAD (`verify.py:446-449`), so it already contains the branch — one `git merge --ff-only <ref>` recovers the whole attempt regardless of which families fired.
- **`_defer` message tail** — in place: ``— attempt work parked at `<ref>`; recover with `git -C "<root>" merge --ff-only <ref>` ``; isolated: ``— failed work kept on branch `<branch>` `` when `scm.keep_failed` (the unit branch is *not* written to `preserve_ref` — a live branch is not a parked snapshot). Empty when nothing was parked, so the notice can never advertise a ref that was not created. The in-place `story-deferred` journal entry gains `preserve_ref=`.
- **Projections** — `documents.py` (`--json`, additive; schema stays 1), `cmd_status` text (`<reason> [<ref>]`), README + FEATURES.

The escalation path is untouched — it does not roll back, so nothing is parked there. No policy or `core.toml` changes.

## Caveat, documented not enforced

`preserve_ref` is reported verbatim from `state.json`. `scm.preserve_keep` prunes the oldest refs at a later run's start, so a named ref may be gone. `status` must not run git, so this is a README/docstring note rather than a validation.

## Tests

Both ablations run before trusting the negative assertions:

1. Delete `task.preserve_ref = None` from the auto-recover arm → `test_rollback_clears_a_previous_attempts_preserve_ref` fails (`assert 'attempt-preserve/run-1-97d596da' is None`).
2. Make the message tail unconditional → `test_defer_notification_stays_bare_when_nothing_was_parked` and `test_defer_recovery_note_is_uniform_across_ref_families` both fail (`merge --ff-only None`).

Also added: both-families subsumption (`git merge-base --is-ancestor branch dirty`), engine E2E asserting the ATTENTION file's ref + command on a budget-exhaustion defer, worktree defer with/without `keep_failed`, model round-trip + legacy-dict default, `status` json + text.

Note on the first draft of ablation 2: the bare-reason case was initially written against a `pre_ready_gate` plugin veto, which turned out to have its own notify site (`engine.py:1002-1003`) and never reaches `_defer` — the test passed under ablation. Rewritten against over-budget sessions, which touch no files, so the exhaustion defer's rollback finds a clean tree and genuinely parks nothing.

`uv run pytest -q -n auto`: 3360 passed, 24 skipped. `trunk check`: clean.

## Summary by CodeRabbit

* **New Features**
  * Deferred notifications now surface the preserved recovery reference (when available) and the corresponding `git ... merge --ff-only` restore guidance, including special wording for worktree isolation and commits-only preservation.
  * `bmad-loop status` (text and `--json`) now includes `preserve_ref` for deferred tasks, and status lines visually annotate it when set.
* **Documentation**
  * Expanded “Plateau-defer” and README worktree failure-recovery docs to explain how preserved refs are formed and reported.
* **Tests**
  * Added/strengthened coverage for `preserve_ref` persistence, clearing, and notification/status output in rollback and worktree scenarios.

### CODE REVIEW

File: src/bmad_loop/recovery_flow.py [hunk 1]
```diff
@@ -332,6 +338,7 @@ def preserve_attempt_commits(self, task: StoryTask, *, allow_pause: bool) -> Non
                 # the operator to blindly `reset --hard` (that would discard them).
                 self.pause_for_manual_recovery(task, baseline, preserve_failed=True)
             return  # re-drive: never pause — proceed to the (human-directed) reset
+        task.preserve_ref = ref
```
[the remark this answers is not in the document]
Comment (pbean) [reply src/bmad_loop/recovery_flow.py:345]: Confirmed and fixed in `89444a6` — thank you, this was real.

Traced the path: `preserve_attempt_commits` assigns the branch, then `preserve_attempt_worktree` catches `verify.GitError` and returns without touching `preserve_ref`, and `safe_reset` runs unconditionally afterwards. `verify.snapshot_worktree` raises at seven points and `_run_git` converts a 120s timeout into one, so it is reachable rather than theoretical.

What settled it was that the PR contradicted itself: `_defer_recovery_note`'s own docstring claimed the note is empty on "a preservation failure", and `model.py` / `recovery_flow.py` both asserted "one `git merge --ff-only <ref>` recovers the whole attempt". Neither held on this path.

One correction to the suggested fix: suppressing the command outright would hide genuinely recoverable committed work, and the ref name alone cannot carry the distinction — a commits-only ref is *complete* when the tree was clean. So the state is recorded rather than derived: a new `StoryTask.preserve_partial` is latched in the `except` arm, and the notice downgrades its claim ("attempt COMMITS parked… the uncommitted changes could not be captured (journal: attempt-worktree-preserve-failed) and did not survive the rollback") while still printing the merge line for the half that survived. `safe_reset` behavior is unchanged.

Regression test added as requested: `test_snapshot_failure_leaves_a_commits_only_ref_flagged_partial` covers exactly the ordering you describe — commits parked, snapshot fails, reset destroys the rest — and asserts the parked ref carries only the committed half. Both ablations were run: deleting the latch fails it, and making the latch unconditional fails the complete-park test.

Whether a worktree-snapshot failure should *block* the reset is a separate question — it collides with #161's "an in-worktree rollback never pauses" invariant and a documented best-effort contract, so it is going to its own issue. This PR fixes the claim, not the policy.

File: README.md [hunk 2]
```diff
@@ -462,7 +462,7 @@ By default work happens **in place** on the checked-out branch (`[scm] isolation
...
+- **`rollback_on_failure`** (default off) — governs in-place recovery when `isolation = "none"`. Off pauses a failed unattended attempt with manual-recovery steps and leaves the tree untouched; on auto-reverts the attempt's tracked changes and removes only the untracked files the run created (never a blanket `git clean`). A resolved escalation's re-drive always auto-recovers regardless of this flag. Under `isolation = "worktree"` the worktree is disposable, so a failed attempt's commits are parked on `attempt-preserve/*` refs and the worktree resets regardless — **`preserve_keep`** (default 20) bounds how many of those recovery refs are kept per family at run start, pruning the oldest (`0` = never prune). Whenever an attempt is parked, the ref is named in the story's defer notification (with the `git merge --ff-only` line that recovers it) and in `bmad-loop status`/`--json` as `preserve_ref` — reported verbatim from run state, so a ref that a later run's `preserve_keep` pruning has since deleted is still shown.
```
[the remark this answers is not in the document]
Comment (pbean) [reply README.md]: Split verdict on this one: the FEATURES.md half is a real defect and is fixed in `89444a6`; the README half I am declining.

**FEATURES.md — agreed.** The sentence enumerated two cases and then attached a singular referent to the union: "in place, *the recovery ref* …; isolated, *the kept-failed unit branch* — and *the ref* is projected as `preserve_ref`". "the ref" has two candidate antecedents across the aside, and the distributive reading is false. Rewritten to scope it explicitly, and to state that the unit branch never is.

**README.md — declining.** Line 465 never mentions `keep_failed`, and its sentence is gated on "Whenever an attempt is **parked**". The two bullets already use disjoint vocabulary for disjoint artifacts: 464 says the kept branch "stay[s] mounted", 465 says parked work goes to recovery refs. I could not construct a reading of 465 that claims the `keep_failed` branch becomes `preserve_ref`.

However, validating this turned up something stronger that the comment inverted. `preserve_ref` is **not** isolation-scoped: `recovery_flow.py` auto-recovers whenever the call happens inside a mounted unit worktree, so a non-fixable dev retry parks on the same shared refs and a deferred isolated unit can carry both a kept-failed branch *and* a real `preserve_ref`. So the risk was not that the branch gets reported as `preserve_ref` — it was that a genuinely parked ref was hidden from the isolated notice while `status --json` displayed it, which is #333's own failure mode. Fixed: the isolated notice now names both, deliberately without a `merge --ff-only` line (that ref is not fast-forwardable there — offering it would land a discarded attempt on the operator's branch).

The PR's guard for this was also inert — `assert task.preserve_ref is None` in a test whose script never rolls back. Its premise is now pinned, and a real isolated-rollback test was added.

(The pre-existing "the worktree resets regardless" on line 465 is indeed loose under `keep_failed`, but it is untouched by this PR — leaving it per the repo's "don't refactor code you aren't touching" doctrine.)

File: src/bmad_loop/engine.py [hunk 3]
```diff
@@ -2999,7 +3024,12 @@ def _defer(self, task: StoryTask, reason: str) -> None:
             # and the worktree kept/dropped by _integrate_unit. Don't touch the
             # tree here (no reset into the main repo — there's nothing to undo).
             self.journal.append("story-deferred", story_key=task.story_key, reason=reason)
```
[the remark this answers is not in the document]
Comment (pbean) [reply src/bmad_loop/engine.py]: The asymmetry is real and is now closed in `89444a6` — but for a different reason than filed, and the stated failure mode does not hold.

**Why the `KeyError` framing does not land.** `journal.append` is `**kwargs` → JSON with no schema, and conditional field presence is the established, documented norm in this journal rather than a defect: `session-end` omits `tokens`/`tokens_weighted` entirely on an aborted end, and README.md says so in prose. `journal.jsonl` is also explicitly not the versioned contract — `machine.py` scopes that to `--json` only. No consumer in `src/` reads `preserve_ref` off a journal entry: `diagnostics.py` counts kinds, the TUI renders `entry.items()` generically, everything else uses `.get()`.

It is also 1-of-3 emit sites, not 1-of-2 — the plugin-veto defer at `engine.py:1002` omits the field too, and it fires in **both** isolation modes. So a hypothetical `entry["preserve_ref"]` consumer would already break on in-place runs, which undercuts the "entries produced by isolated runs" characterisation.

**What was actually wrong.** Validating this surfaced the real issue underneath: `preserve_ref` is not isolation-scoped. A unit worktree's own dev-retry rollback parks on the same shared refs, so a deferred isolated unit can carry a genuine ref — which the isolated notice never named even though `status --json` reported it. That is #333's own failure mode, reproduced in the isolated case, and it made `model.py`'s "None = … or an isolated unit" comment false.

Fixing that carries your suggestion with it: the isolated arm now emits `preserve_ref=task.preserve_ref or ""`, matching the in-place arm, so the field is present on both `_defer` arms. Adopted on the symmetry argument, not the `KeyError` one.

File: src/bmad_loop/engine.py [hunk 4]
```diff
@@ -3018,12 +3048,17 @@ def _defer(self, task: StoryTask, reason: str) -> None:
                 if current != snapshot:
                     deferred_work.parent.mkdir(parents=True, exist_ok=True)
                     deferred_work.write_text(snapshot, encoding="utf-8")
-        self.journal.append("story-deferred", story_key=task.story_key, reason=reason)
+        self.journal.append(
+            "story-deferred",
+            story_key=task.story_key,
+            reason=reason,
+            preserve_ref=task.preserve_ref or "",
```
[the remark this answers is not in the document]
Comment (pbean) [reply src/bmad_loop/engine.py:3082]: Accurate observation, but declining: this follows the local convention rather than breaking one.

`recovery_flow.py` already writes `baseline=task.baseline_commit or ""` on the `rollback-auto` entry — emitted two lines before the ref is parked, in the same recovery cluster this PR extends, for a field that is likewise `str | None` in the model and serializes as `null` in `state.json`. So the `""`-in-journal / `null`-in-state divergence you describe already exists on the immediately adjacent entry; the new line is consistent with its neighbours. (Both conventions do exist in this journal — `session-end` passes literal `None` for tokens — but `or ""` is the one used here.)

On impact: nothing joins the two representations. No consumer in `src/` reads `preserve_ref` off a journal entry at all, so the only observable difference is cosmetic — a bare `preserve_ref=` in the TUI journal row and `""` instead of `null` in a diagnostics dump, both advisory surfaces.

Worth noting the two findings in this review are in mild tension: this one wants the field to carry `None`, while the sibling comment wants it always present. Satisfying both would mean emitting `preserve_ref=None` at all three `story-deferred` sites — a different convention from the one the surrounding rollback/preserve code uses. I have taken the presence half (the isolated arm now emits the field for symmetry) and left the sentinel as `""` to match `rollback-auto`.

The genuinely useful thing this review surfaced is on the sibling thread: `preserve_ref` is not isolation-scoped, which made a comment in `model.py` false and hid a real ref from the isolated defer notice. That is fixed in `89444a6`.
